src/lib/operators/aggregate.cpp

```diff
@@ -702,15 +691,43 @@ std::enable_if_t<func == AggregateFunction::Avg && !std::is_arithmetic_v<Aggrega
 void Aggregate::_write_groupby_output(PosList& pos_list) {
   auto input_table = input_table_left();
 
-  for (size_t group_column_index = 0; group_column_index < _groupby_column_ids.size(); ++group_column_index) {
-    auto base_segments = std::vector<std::shared_ptr<const BaseSegment>>();
-    for (const auto& chunk : input_table->chunks()) {
-      base_segments.push_back(chunk->get_segment(_groupby_column_ids[group_column_index]));
-    }
-    _groupby_segments[group_column_index]->reserve(pos_list.size());
-    for (const auto& row_id : pos_list) {
-      _groupby_segments[group_column_index]->append((*base_segments[row_id.chunk_id])[row_id.chunk_offset]);
-    }
+  // For each GROUP BY column, resolve its type, iterate over its values, and add them to a new output ValueSegment
+  for (const auto& column_id : _groupby_column_ids) {
+    _output_column_definitions.emplace_back(input_table->column_name(column_id),
+                                            input_table->column_data_type(column_id));
+
+    resolve_data_type(input_table->column_data_type(column_id), [&](const auto typed_value) {
+      using ColumnDataType = typename decltype(typed_value)::type;
+
+      auto values = pmr_concurrent_vector<ColumnDataType>(pos_list.size());
+      auto null_values = pmr_concurrent_vector<bool>(pos_list.size());
+      std::vector<std::unique_ptr<BaseSegmentAccessor<ColumnDataType>>> accessors(input_table->chunk_count());
+
+      auto output_offset = ChunkOffset{0};
+
+      for (const auto& row_id : pos_list) {
+        // pos_list was generated as part of the aggregate operator. While it might point to rows that contain NULL
+        // values, no new NULL values should have been added.
+        DebugAssert(!row_id.is_null(), "Did not expect NULL value here");
+
+        auto& accessor = accessors[row_id.chunk_id];
+        if (!accessor) {
+          accessor =
+              create_segment_accessor<ColumnDataType>(input_table->get_chunk(row_id.chunk_id)->get_segment(column_id));
+        }
+
+        const auto& optional_value = accessor->access(row_id.chunk_offset);
+        if (!optional_value) {
+          null_values[output_offset] = true;
```

Comment:
Can we get rid of the branch?

`null_values[output_offset] = optional_value;`

Reply:
We would still need the branch because `*opt` is undefined if `opt` is empty. Also, it would hide our intent.

Reply:
Well, a comment would sufficiently cover our intend and if the branch could be removed, I'd at least try it.

So something (forgive my C++ knowledge) like `null_values[output_offset] = static_cast<bool>(optional_value);` does not work?

Reply:
But how would you write to `values`?

Reply:
Ok, I see. Doing that would be some really nasty branchless code. Not worth it.